### PR TITLE
Remote connection race condition

### PR DESCRIPTION
**Pull request by alanmcgovern**
This fixes one race condition and then rewrites some continuation based code to be async/await as it is a bit easier to reason about.

### CODE REVIEW

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core.Execution/RemoteProcessConnection.cs [hunk 1]
```diff
@@ -530,7 +530,8 @@ void OnConnected (IAsyncResult res)
 
 		public Task ProcessPendingMessages ()
 		{
-			return Task.WhenAll (pendingMessageTasks.ToArray ());
+			lock (pendingMessageTasks)
+				return Task.WhenAll (pendingMessageTasks.ToArray ());
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Core.Execution/RemoteProcessConnection.cs]: 👍

Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Core.Execution/RemoteProcessConnection.cs]: Or maybe this is holding the lock a bit too much actually, until it returns a task from .WhenAll(). Lock contention should be avoided where possible, so I'd go with the more verbose way of only locking the ToArray() call.

Comment (alanmcgovern) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Core.Execution/RemoteProcessConnection.cs]: I think it's ok to leave it as is as the extra call is super fast. If you really want it changed we can do that :)

Comment (alanmcgovern) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Core.Execution/RemoteProcessConnection.cs]: I'll just merge it. It's much easier to read as-is and will never be a perf bottleneck :)

**Pull request merged**